docs/book/src/advanced/generics_and_trait_constraints.md

```diff
@@ -0,0 +1,125 @@
+# Generics and Trait Constraints
+
+## Generics as Constraints
+
+At a high level, Sway allows you to define constraints, or restrictions, that
+allow you to strike a balance between writing abstract and reusable code and
+enforcing compile-time checks to determine if the abstract code that you've
+written is correct.
+
+The "abstract and reusable" part largely comes from generic types and the
+"enforcing compile-time checks" part largely comes from trait constraints.
+Generic types can be used with functions, structs, and enums (as we have seen in
```

Comment:
When mentioning another part of the book it's useful to link that part so that the reader does not have to search and navigate to it themselves (even if it is the previous section).